### PR TITLE
account settings extension fields for Ficus

### PR DESCRIPTION
**Pull request by bryanlandia**
This brings the Dogwood feature up to date for Ficus

### CODE REVIEW

**Comment by melvinsoft** [comment]
@bryanlandia Just a general question, isn't this overlapping with the already built in functionality in the platform that supports https://github.com/appsembler/custom-form-app ?

**Comment by bryanlandia** [comment]
@melvinsoft - No, those are Registration form extension fields.  What this does is allow you to specify custom fields to show on the Account Settings page `/account/`.  They are useful together, because if you create a custom field for registration it is nice to make it viewable and editable in Account Settings.  You point to objects describing  new fields in the same way as for the registration extension form in config--by dotted path. It was a kind of complicated change to make the Backbone.js and py views pull in the new fields.  

Do you still want to review this one?

**Comment by bryanlandia** [comment]
@melvinsoft @tkeemon - Can I get another look at this one soon?  ISC, ISC CMC, and pretty soon ExtraCare are all gonna want this merged to ficus/master rather than relying on custom branches.

File: lms/djangoapps/student_account/fields.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+"""
+Example AccountSettingsExtensionField subclass:
+the below wouldn't be defined here, but in some other module
+or add on app
+
+
+class BryanTestExtensionField(AccountSettingsExtensionField):
```
Comment (melvinsoft) [inline lms/djangoapps/student_account/fields.py]: Should this class be re removed? Since is a test.

**Comment by abeals** [comment]
@johnbaldwin I also added you to this pull request because it overlaps *heavily* with what you've been thinking through for custom account fields for reporting. **Please** review this PR and familiarize yourself with the approach Bryan is taking here.

**Comment by bryanlandia** [comment]
@johnbaldwin I'm going to call in that PR review trade :)  Please take a look at this.  I'd like to get it merged.

File: lms/static/js/student_account/views/account_settings_factory.js [hunk 2]
```diff
@@ -268,16 +332,15 @@
...
-                        success: showAccountSettingsPage,
-                        error: showLoadingError
-                    });
-                },
-                error: showLoadingError
-            });
+
+            fetchAccountExtensionModels = function() {
+                //fetch each of the extension field models
+                // TODO: is it important to chain these as on success callbacks?
```
Comment (johnbaldwin) [inline lms/static/js/student_account/views/account_settings_factory.js]: @bryanlandia Does this need to be chained?

**Comment by johnbaldwin** [comment]
My backbone.js is still pretty weak, so I'm likely to miss issues, especially subtle ones.

Have you tested this in a staging instance?

File: lms/djangoapps/student_account/views.py [hunk 3]
```diff
@@ -49,9 +53,16 @@
...
+from openedx.core.djangoapps.user_api.accounts.api import request_password_change
+from openedx.core.djangoapps.user_api.errors import UserNotFound
+
+from student_account.fields import AccountSettingsExtensionField
+
+
 AUDIT_LOG = logging.getLogger("audit")
 log = logging.getLogger(__name__)
 User = get_user_model()  # pylint:disable=invalid-name
+ERR_LOG = logging.getLogger(__name__)
```
Comment (johnbaldwin) [inline lms/djangoapps/student_account/views.py]: @bryanlandia How is this different from the log object two lines above?

File: lms/djangoapps/student_account/views.py [hunk 4]
```diff
@@ -18,6 +19,9 @@
 from django.views.decorators.csrf import ensure_csrf_cookie
 from django.views.decorators.http import require_http_methods
 from django_countries import countries
+
+from django.core.exceptions import ImproperlyConfigured
```
Comment (johnbaldwin) [inline lms/djangoapps/student_account/views.py]: small nit: Please move up above `django_countries` with the django package imports

File: lms/djangoapps/student_account/views.py [hunk 5]
```diff
@@ -49,9 +53,16 @@
 from util.date_utils import strftime_localized
 from util.enterprise_helpers import set_enterprise_branding_filter_param
 
+from openedx.core.djangoapps.user_api.accounts.api import request_password_change
+from openedx.core.djangoapps.user_api.errors import UserNotFound
```
Comment (johnbaldwin) [inline lms/djangoapps/student_account/views.py]: Another small nit: Please move these two to be with the other openedx.core imports

**Comment by bryanlandia** [comment]
@johnbaldwin Yes, it's deployed to two staging sites, InterSystems' main site and CMC site.  The Eucalyptus version is also in production for Trinity and ExtraCare.

File: lms/djangoapps/student_account/views.py [hunk 6]
```diff
@@ -499,3 +514,45 @@ def account_settings_context(request):
         } for state in auth_states if state.provider.display_for_login or state.has_account]
 
     return context
+
+
+def get_account_settings_extension_config(request):
```
Comment (johnbaldwin) [inline lms/djangoapps/student_account/views.py]: If you reasonably can, I really urge you to to pull the business logic out of the view and into its own module

File: lms/static/js/student_account/views/account_settings_factory.js [hunk 7]
```diff
@@ -1,16 +1,29 @@
 (function(define, undefined) {
     'use strict';
+
+    define('extension_deps', ['underscore'], function(_) {
```
[the remark this answers is not in the document]
Comment (bryanlandia) [reply lms/static/js/student_account/views/account_settings_factory.js]: @johnbaldwin - No, I'm just using Require here and defining a local function within this Backbone factory.  The top level function defines the factory but within it it's ok to define other functions.

File: lms/static/js/student_account/views/account_settings_factory.js [hunk 8]
```diff
@@ -241,7 +256,56 @@
...
             ];
 
-            accountSettingsView = new AccountSettingsView({
+            // extension fields
+            var deps, ext_fields;
+            var ext_deps_config = extension_deps(extensionFieldsData);
+
+            // http://stackoverflow.com/a/17448869
+            RequireJS.require(_.values(ext_deps_config), function() {
+                // TODO: some defensive type checking
```
[the remark this answers is not in the document]
Comment (bryanlandia) [reply lms/static/js/student_account/views/account_settings_factory.js]: ah, well looks like it.  I'm afraid my head's not in it enough any more to go do that very efficiently right now.

File: lms/djangoapps/student_account/views.py (hunk 3 above)
Comment (bryanlandia) [reply lms/djangoapps/student_account/views.py]: @johnbaldwin The `AUDIT_LOG` was from the original code, and seems to be more about auditing for security around password attempts and resets.  I wanted to use a separate log for application errors.

File: lms/djangoapps/student_account/views.py (hunk 5 above)
Comment (bryanlandia) [reply lms/djangoapps/student_account/views.py]: These were actually duplicates.  I removed them.

File: lms/djangoapps/student_account/views.py (hunk 6 above)
Comment (bryanlandia) [reply lms/djangoapps/student_account/views.py]: I'm going to keep it how it is for now, as the views module here includes a number of similar methods for getting context for the view.

File: lms/djangoapps/student_account/views.py (hunk 4 above)
Comment (bryanlandia) [reply lms/djangoapps/student_account/views.py]: Done. Thanks

File: lms/static/js/student_account/views/account_settings_factory.js (hunk 2 above)
Comment (bryanlandia) [reply lms/static/js/student_account/views/account_settings_factory.js]: I should have already removed this TODO.

File: lms/static/js/student_account/views/account_settings_factory.js [hunk 9]
```diff
@@ -241,7 +256,56 @@
...
+                deps = _.object(_.keys(ext_deps_config), arguments);
+                ext_fields = _.map(extensionFieldsData, function(extfield) {
+                    var model_inst, field_view_class;
+                    model_inst = new deps[extfield.id]();
+                    model_inst.url = extfield.api_url;
+                    field_view_class = eval(extfield.js_field_view_class);
+                    return {
+                        'view': new field_view_class({ // TODO: determine which view
```
[the remark this answers is not in the document]
Comment (bryanlandia) [reply lms/static/js/student_account/views/account_settings_factory.js]: I really don't remember what this TODO was about.  I think it's good as is, so I'm removing it.

File: lms/djangoapps/student_account/fields.py (hunk 1 above)
Comment (bryanlandia) [reply lms/djangoapps/student_account/fields.py]: @melvinsoft This is part of a docstring that is essentially documentation of how to set up a subclass of AccountSettingsExtensionField, but I'll rename it.

**Pull request merged**